libs/ledger-live-common/src/hw/installLanguage.ts

```diff
@@ -0,0 +1,189 @@
+
+              const pack = packs[0];
+
+              const { apdu_install_url } = pack;
+              const url = apdu_install_url;
+
+              const { data: rawApdus } = await network({
+                method: "GET",
+                url,
+              });
+
+              const apdus = rawApdus.split(/\r?\n/).filter(Boolean);
+
+              await uninstallAllLanguages(transport);
+
+              for (let i = 0; i < apdus.length; i++) {
+                if (apdus[i].startsWith("e030")) {
+                  subscriber.next({
+                    type: "devicePermissionRequested",
+                  });
+                }
+
+                const response = await transport.exchange(Buffer.from(apdus[i], "hex"));
+                const status = response.readUInt16BE(response.length - 2);
+                const statusStr = status.toString(16);
+
+                // Some error handling
+                if (status === 0x5501) {
+                  return subscriber.error(new LanguageInstallRefusedOnDevice(statusStr));
+                } else if (status !== 0x9000) {
+                  return subscriber.error(new TransportError("Unexpected device response", statusStr));
+                }
+
+                subscriber.next({
+                  type: "progress",
+                  progress: (i + 1) / apdus.length,
+                });
+              }
+
+              subscriber.next({
+                type: "languageInstalled",
+              });
+
+              subscriber.complete();
+            }),
+            catchError((e: Error) => {
+              if (
+                e instanceof DeviceOnDashboardExpected ||
+                (e &&
+                  e instanceof TransportStatusError &&
+                  [0x6e00, 0x6d00, 0x6e01, 0x6d01, 0x6d02].includes(
+                    // @ts-expect-error typescript not checking agains the instanceof
+                    e.statusCode
+                  ))
+              ) {
+                const quitAppObservable = from(getAppAndVersion(transport)).pipe(
+                  concatMap((appAndVersion) => {
+                    return !isDashboardName(appAndVersion.name)
+                      ? attemptToQuitApp(transport, appAndVersion)
+                      : of<InstallLanguageEvent>({
+                          type: "appDetected",
+                        });
+                  })
+                );
+
+                quitAppObservable.subscribe(
+                  (event) => subscriber.next(event),
+                  (error) => subscriber.error(error)
+                );
+              }
+
+              subscriber.error(e);
+              return EMPTY;
+            })
+          )
+          .subscribe();
+
+        return () => {
+          timeoutSub.unsubscribe();
+          sub.unsubscribe();
+        };
+      })
+  );
+
+  return sub;
+}
+
+const uninstallAllLanguages = async (transport: Transport) => {
+  // TODO: in a future FW version, this will be a single apdu
```

Comment:
Didn't they add it already with `e033ff00` or whatever it was? It's crazy we are still doing this loop from the poc

Reply:
not yet. It's coming